Use structured JSON for interfacing with 'ClientRequest' / 'ClientResponse'

KtorZ opened the pull request:
- :round_pushpin: **Use structured JSON for interfacing with 'ClientRequest'**

- :round_pushpin: **Remove use of aeson-casing**
    Instead, favor hand-written JSON instances over generic with options.
  The rationale being that generic deriving with options and other
  meta-programming are often hard to read and easy to shoot oneself in
  the foot.

  Instead, using JSON builder makes for transparent definitions at the
  cost of a small maintainance burden when changing data-types.

- :round_pushpin: **Replace use of aesonQQ with builders**
    QuasiQuoters don't make unanimity, but builders do despite being
  perhaps a bit _too polymorphic_.

- :round_pushpin: **Write JSON roundtrip properties for `ClientRequest`**
    This was slightly more annoying than I initially thought because of
  the shrinkers. I thought I could get away with `genericShrink` and a
  `Generic` instance on `ClientRequest`, but this requires the subtypes
  to have `Arbitrary` instances, which is pretty annoying since we're
  trying to precisely avoid those.

  Ended up writing the shrinker by hand, not sure if it's worth it given
  that the model is small enough in practice anyway...

- :round_pushpin: **Write JSON instances for 'ClientResponse'**

- :round_pushpin: **Add {From,To}JSON / {From,To}CBOR / Arbitrary instances to hydra-prelude**
    All used pervasively across the codebase. So good candidates to be factored out in the prelude.

- :round_pushpin: **Define Arbitrary instances for base data-types in lib code**
    After discussion, we concluded that it is pretty useful to have
  generic Arbitrary instances readily available for base data-types, and
  only provide hand-written generators on a case-by-case. This leverages
  the generic-random package for generating arbitrary instances, and
  quickcheck for shrinkers when possible.

- :round_pushpin: **Refactor End-to-End to use smart-constructors for building JSON requests.**

- :round_pushpin: **Reply with structured JSON instead of mere 'show'**

- :round_pushpin: **Rework 'waitForResponse' expectation failure message**
    Enhance readability, as a big unformatted show is like looking for a needle in a haystack. Not good for quick debugging.

- :round_pushpin: **Represent 'Party' as JSON integers**
    The 'real' implementation should likely be more like it was before
  this commit, a base16 or bech32 encoding of the key. But with the mock
  cryptos, keys are basically just numbers and it's easier to read than
  long bytestring in end-to-end tests. Arguably, the end-to-end
  interface should not change when we change the crypto, so this may be
  only midly correct...

- :round_pushpin: **Rename 'ClientRequest' -> 'ClientInput', 'ClientResponse' -> 'ServerOutput'**

KtorZ committed:
Use structured JSON for interfacing with 'ClientRequest'

ch1bo commented on local-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -52,16 +57,16 @@ spec = describe "End-to-end test using a mocked chain though" $ do
-                sendRequest n3 "Commit (fromList [3])"
+                sendRequest n1 [aesonQQ|{"req": "commit", "args": [1] }|]
+                sendRequest n2 [aesonQQ|{"req": "commit", "args": [2] }|]
+                sendRequest n3 [aesonQQ|{"req": "commit", "args": [3] }|]
```
Why do we need quasi quotes here?

KtorZ replied:
Because it's handy and readable for constructing JSON ?

KtorZ replied:
(if you have a better proposal, I am glad to hear it :no_mouth:)

ch1bo replied:
I have used "builders" for an `Aeson.Value` in the past, but have also seem full string quasi-quotation. This seems to be something in between from the looks of it. Personally I prefer to use the former as the compiler helps in finding bugs when creating the json AST.

The following should be equal, but I would be interested in seeing how the two approaches compare in a more complicated example:

```suggestion
                sendRequest n3 $ object [ "req" .= "commit", "args" .= [3] ]
```

KtorZ replied:
>  I don't see what  QQuoters add here, it adds a dependency and compillation time for little value.

Well, actually, the quoter comes from the aeson library itself so it's not like it is adding much dependency. Plus, it's not a mere string QQ, it does parse the JSON underneath and provides compile-time errors for invalid JSON. The added value to me really being in the "what you see is what you get" kind of spirit. With builders and implicit conversions it's sometimes easy to miss something.

ch1bo commented on hydra-node/src/Hydra/Ledger/Simple.hs:
```diff
@@ -27,6 +35,12 @@ data SimpleTx = SimpleTx
   deriving stock (Eq, Ord, Generic, Read, Show)
 
+instance ToJSON SimpleTx where
+  toJSON = genericToJSON (aesonPrefix camelCase)
```
So this is creating a JSON object with fields `id`, `inputs` and `outputs`? IMO this "drop all lowercase letters until first uppercase" convention of this `aesonPrefix` is only sligthly better than `drop 2` and would like to avoid such things as much as possible. Can we stick to just use field names as is and make then non-overlapping or non-conflicting on a case by case basis?

KtorZ replied:
> Can we stick to just use field names as is and make then non-overlapping or non-conflicting on a case by case basis?

I am fine with that ^.^

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -50,6 +62,25 @@ data ClientRequest tx
+clientRequestJsonOptions :: Aeson.Options
+clientRequestJsonOptions =
+  Aeson.defaultOptions
+    { constructorTagModifier = camelCase
```
This `camelCase` function is basically `mapHead toLower` and I know now why you were suggesting to avoid `UTxO` as a name as it would turn out as `uTxO` using this constructor modifier :joy:

Not sure though whether we want to have another dependency just for this very simple modifier

ch1bo replied:
Instead of adding dependencies and doing fancy meta-programming I'd be even open to hand-written instances. Sometimes they are easier to read and maintain than `genericXXX` + various `options` and `modifiers`

KtorZ replied:
> Not sure though whether we want to have another dependency just for this very simple modifier

I initially wrote them by hand, and ended up writing the same one in two places. Wanted to move in a separate module but it felt a bit _too excessive_ / inappropriate. In the end, aeson-casing is a tiny library with almost no dependency (base and aeson) which kinda does just is needed. 

Writing JSON by hand is always doable, just a bit annoying. I think my favorite approach remains generics with default options, and possibly `DuplicateRecordFields` when really necessary. With `NamedFieldPuns` it's usually not a big problem.

ch1bo reviewed (changes requested):
The `aesonPrefix` kinda rubs me the wrong way (have been bitten by `drop 16` as a field modifier in the past) and would be interested in alternatives.

KtorZ commented on hydra-node/src/Hydra/API/Server.hs:
```diff
@@ -24,12 +25,12 @@ data APIServerLog
 
 withAPIServer ::
-  Tx tx =>
+  (Tx tx, FromJSON tx, FromJSON (UTxO tx)) =>
```
Thought of it, and even considered maybe to pass codecs as a record object actually, to not necessarily tied to JSON all-the-way. Problem is that we'll likely have to have JSON instances anyway so.. the benefits of abstracting away the codec is maybe not that great. 

Moving it to `Tx` is a possibility :+1:

KtorZ commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -50,6 +62,25 @@ data ClientRequest tx
+          { tagFieldName = "req"
+          , contentsFieldName = "args"
+          }
+    }
```
No. But the default names from aeson are `tag` and `content` which I very much dislike. So I went for something a bit more appropriate IMO.

KtorZ committed:
Remove use of aeson-casing

  Instead, favor hand-written JSON instances over generic with options.
  The rationale being that generic deriving with options and other
  meta-programming are often hard to read and easy to shoot oneself in
  the foot.

  Instead, using JSON builder makes for transparent definitions at the
  cost of a small maintainance burden when changing data-types.

KtorZ committed:
Replace use of aesonQQ with builders

  QuasiQuoters don't make unanimity, but builders do despite being
  perhaps a bit _too polymorphic_.

KtorZ committed:
Write JSON roundtrip properties for `ClientRequest`

  This was slightly more annoying than I initially thought because of
  the shrinkers. I thought I could get away with `genericShrink` and a
  `Generic` instance on `ClientRequest`, but this requires the subtypes
  to have `Arbitrary` instances, which is pretty annoying since we're
  trying to precisely avoid those.

  Ended up writing the shrinker by hand, not sure if it's worth it given
  that the model is small enough in practice anyway...

KtorZ committed:
Write JSON instances for 'ClientResponse'

KtorZ committed:
Add {From,To}JSON / {From,To}CBOR / Arbitrary instances to hydra-prelude

  All used pervasively across the codebase. So good candidates to be
  factored out in the prelude.

KtorZ committed:
Define Arbitrary instances for base data-types in lib code

  After discussion, we concluded that it is pretty useful to have
  generic Arbitrary instances readily available for base data-types, and
  only provide hand-written generators on a case-by-case. This leverages
  the generic-random package for generating arbitrary instances, and
  quickcheck for shrinkers when possible.

KtorZ committed:
Refactor End-to-End to use smart-constructors for building JSON requests.

KtorZ committed:
Represent 'Party' as JSON integers

  The 'real' implementation should likely be more like it was before
  this commit, a base16 or bech32 encoding of the key. But with the mock
  cryptos, keys are basically just numbers and it's easier to read than
  long bytestring in end-to-end tests. Arguably, the end-to-end
  interface should not change when we change the crypto, so this may be
  only midly correct...

KtorZ committed:
Reply with structured JSON instead of mere 'show'

KtorZ committed:
Rework 'waitForResponse' expectation failure message

  Enhance readability, as a big unformatted show is like looking for a needle in a haystack. Not good for quick debugging.

KtorZ committed:
Rename 'ClientRequest' -> 'ClientInput', 'ClientResponse' -> 'ServerOutput'

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -57,12 +49,26 @@ data ClientRequest tx
+
+  -- NOTE: Somehow, can't use 'genericShrink' here as GHC is complaining about
+  -- Overlapping instances with 'UTxO tx' even though for a fixed `tx`, there
+  -- should be only one 'UTxO tx'
```
Maybe we're missing a functional dependency or so?

KtorZ replied:
Thought so. But really, it shouldn't be necessary here. I didn't give it much thought.
